replica/table.cc

```diff
@@ -580,49 +580,90 @@ table::seal_active_memtable(flush_permit&& flush_permit) noexcept {
+                _memtables->add_memtable();
+                st = state::start_op;
+                _stats.memtable_switch_count++;
+                // This will set evictable occupancy of the old memtable region to zero, so that
+                // this region is considered last for flushing by dirty_memory_manager::flush_when_needed().
+                // If we don't do that, the flusher may keep picking up this memtable list for flushing after
+                // the permit is released even though there is not much to flush in the active memtable of this list.
+                old->region().ground_evictable_occupancy();
+                memtable_size = old->occupancy().total_space();
+                [[fallthrough]];
+            case state::start_op:
+                previous_flush = _flush_barrier.advance_and_await();
+                op = _flush_barrier.start();
+                st = state::try_flush;
+                _stats.pending_flushes++;
+                _config.cf_stats->pending_memtables_flushes_count++;
+                _config.cf_stats->pending_memtables_flushes_bytes += memtable_size;
+                [[fallthrough]];
+            case state::try_flush: {
+                // Reacquiring the write permit might be needed if retrying flush
+                if (!permit.has_sstable_write_permit()) {
+                    tlogger.debug("seal_active_memtable: reacquiring write permit");
+                    permit = co_await std::move(permit).reacquire_sstable_write_permit();
+                }
+                auto write_permit = permit.release_sstable_write_permit();
+            // FIXME: badly indented on purpose
+            if (co_await this->try_flush_memtable_to_sstable(old, std::move(write_permit)) == stop_iteration::yes) {
+                st = state::done;
+                r.reset();
+                // break out of the retry loop
+                continue;
+            }
+            break;
+            }
+            case state::done:
+                __builtin_unreachable();
+            }
+        } catch (...) {
+            e = std::current_exception();
+            _config.cf_stats->failed_memtables_flushes_count++;
+            try {
+                std::rethrow_exception(e);
+            } catch (const std::bad_alloc& e) {
+                // There is a chance something else will free the memory, so we can try again
+            } catch (...) {
+                try {
+                    // At this point we don't know what has happened and it's better to potentially
+                    // take the node down and rely on commitlog to replay.
+                    on_internal_error(tlogger, e);
+                } catch (const std::exception& ex) {
+                    // If the node is configured to not abort on internal error,
+                    // but propagate it up the chain, we can't do anything reasonable
+                    // at this point. The error is logged and we can try again later
+                }
```

Comment:
This means we don't take the node down. Do we or don't we?

Reply:
Ah, it's a preexisting problem (later solved).

Reply:
Added FIXME comment in v10